File: FruitSaladAi/src/ai/Ai.java
/**
 * AI - This AI communicates with the game to do some unit testing on the
 * platform.
 */

package ai;

import game.FruitSaladAi;
import game.Api;

public class Ai implements FruitSaladAi {
	// -------------------------------------------------------------------------
	// Public methods
	// -------------------------------------------------------------------------
	
	@Override
	public void initGame(int[][] architecture, int[][] fruits,
		int[][] buildings, int limitCherry, int limitKiwi, int limitNut) {
		System.out.println("AI: Initialize...");
		
		// Display architecture
		System.out.println("== Architecture ==");
		System.out.println("{");
		for (int i = 0; i < architecture.length; i++) {
			System.out.print("\t{");
			for (int j = 0; j < architecture[0].length; j++) {
				System.out.print(architecture[i][j] + ", ");
			}
			System.out.println("},");
		}
		System.out.println("}");
		System.out.println();
		
		// Display fruits
		System.out.println("== Fruits ==");
		for (int i = 0; i < fruits.length; i++) {
			System.out.println("Fruit: "
				+ Api.decode(fruits[i][Api.OBJECT_TYPE]) + " " + "id #"
				+ fruits[i][Api.OBJECT_ID] + " at (" + fruits[i][Api.OBJECT_X]
				+ ";" + fruits[i][Api.OBJECT_Y] + ")");
		}
		System.out.println("Limits: " + limitCherry + " cherries, " + limitKiwi
			+ " kiwies, " + limitNut + " nuts");
		System.out.println();
		
		// Display buildings
		System.out.println("== Buildings ==");
		for (int i = 0; i < buildings.length; i++) {
			System.out.println("Building: "
				+ Api.decode(buildings[i][Api.OBJECT_TYPE]) + " " + "id #"
				+ buildings[i][Api.OBJECT_ID] + " at ("
				+ buildings[i][Api.OBJECT_X] + ";" + buildings[i][Api.OBJECT_Y]
				+ ")");
		}
		System.out.println();
		
		// Test game API functions
		testGameApi();
	}
	
	@Override
	public void playTurn(int[][] newObjects, int[] deletedObjects,
		int[][] movedFruits, int[][] modifiedFruits, int[][] modifiedSugarDrops) {
		System.out.println("AI: Play...");
		
		// Display new objects
		System.out.println("== New objects ==");
		for (int i = 0; i < newObjects.length; i++) {
			System.out.println("New object: "
				+ Api.decode(newObjects[i][Api.OBJECT_TYPE]) + " " + "id #"
				+ newObjects[i][Api.OBJECT_ID] + " at ("
				+ newObjects[i][Api.OBJECT_X] + ";"
				+ newObjects[i][Api.OBJECT_Y] + ") ; info="
				+ newObjects[i][Api.OBJECT_INFO]);
		}
		System.out.println();
		
		// Display deleted objects
		System.out.println("== Deleted objects ==");
		for (int i = 0; i < deletedObjects.length; i++) {
			System.out.println("Object: id #" + deletedObjects[i] + " deleted");
		}
		System.out.println();
		
		// Display moved fruits objects
		System.out.println("== Moved fruits ==");
		for (int i = 0; i < movedFruits.length; i++) {
			System.out.println("Fruit: id #" + movedFruits[i][Api.OBJECT_ID]
				+ " moved from (" + movedFruits[i][Api.OBJECT_X] + ";"
				+ movedFruits[i][Api.OBJECT_Y] + ") to ("
				+ movedFruits[i][Api.OBJECT_NEW_X] + ";"
				+ movedFruits[i][Api.OBJECT_NEW_Y] + ")");
		}
		System.out.println();
		
		// Display modified fruits objects
		System.out.println("== Modified fruits ==");
		for (int i = 0; i < modifiedFruits.length; i++) {
			System.out.println("Fruit: id #" + modifiedFruits[i][Api.OBJECT_ID]
				+ " now has " + modifiedFruits[i][Api.OBJECT_LIFE] + " PV and "
				+ modifiedFruits[i][Api.OBJECT_DEFENSE] + " Def");
		}
		System.out.println();
		
		// Display modified sugar drops objects
		System.out.println("== Modified sugar drops ==");
		for (int i = 0; i < modifiedSugarDrops.length; i++) {
			System.out.println("Sugar drop: id #"
				+ modifiedSugarDrops[i][Api.OBJECT_ID] + " now has "
				+ modifiedSugarDrops[i][Api.OBJECT_SUGAR] + " sugar grains");
		}
		System.out.println();
	}
	
	@Override
	public void chestOpened(int chestId, int[][] equipments) {
		System.out.println("AI: Chest #" + chestId + " opened...");
		
		// Display chest equipments
		System.out.println("== Chest equipments ==");
		for (int i = 0; i < equipments.length; i++) {
			System.out.println("Equipment: "
				+ Api.decode(equipments[i][Api.OBJECT_TYPE]) + " id #"
				+ equipments[i][Api.OBJECT_ID] + " at ("
				+ equipments[i][Api.OBJECT_X] + ";"
				+ equipments[i][Api.OBJECT_Y] + ")");
		}
		System.out.println();
	}
	
	@Override
	public void stop() {
		System.out.println("AI: Stopped.");
	}
	
	// -------------------------------------------------------------------------
	// Private methods
	// -------------------------------------------------------------------------
	
	private void testGameApi() {
		System.out.println("move: " + Api.decode(Api.move(1, 42, 42)));
		
		System.out.println("attack: " + Api.decode(Api.attack(1, 2)));
		
		System.out.println("useEquipment: "
			+ Api.decode(Api.useEquipment(1, 512, 1)));
		
		System.out.println("pickUpEquipment: "
			+ Api.decode(Api.pickUpEquipment(1, 57)));
		
		System.out.println("dropEquipment: "
			+ Api.decode(Api.dropEquipment(1, 72, 5, 6)));
		
		System.out.println("dropSugar: "
			+ Api.decode(Api.dropSugar(1, 10, 5, 6)));
		
		System.out.println("openChest: " + Api.decode(Api.openChest(1, 18)));
		
		System.out.println("stockSugar: " + Api.decode(Api.stockSugar(1)));
		
		System.out.println("sellEquipment: "
			+ Api.decode(Api.sellEquipment(1, 16)));
		
		System.out.println("buyEquipment: "
			+ Api.decode(Api.buyEquipment(1, Api.EQUIPMENT_SALT_SNIPER)));
		
		System.out.println("drinkJuice: " + Api.decode(Api.drinkJuice(1)));
		
		System.out.println("fructify: "
			+ Api.decode(Api.fructify(1, Api.FRUIT_CHERRY, 1, 4)));
		
		System.out.println("drawVitamin: " + Api.decode(Api.drawVitamin(1)));
		
		System.out.println("writeText: "
			+ Api.decode(Api.writeText("Hello frutty world")));
		
		System.out.println("writeTextAt: "
			+ Api.decode(Api.writeTextAt("Hello from 2;2", 2, 2)));
		
		System.out.println("drawLine: "
			+ Api.decode(Api.drawLine(1, 1, 5, 8, Api.COLOR_RED)));
		
		System.out.println("drawCircle: "
			+ Api.decode(Api.drawCircle(8, 9, 2, Api.COLOR_BLUE)));
		
		System.out.println("colorSquare: "
			+ Api.decode(Api.colorSquare(6, 2, Api.COLOR_YELLOW)));
	}
}

File: COW/games/FruitSalad/ais/VitaminCpp/src/Ai.cpp
#include "Ai.hpp"
#include "Variant.hpp"
#include <iostream>
#include <sys/time.h>
#include <unistd.h>
#include <chrono>

using namespace std;

Ai ai = Ai();

Ai::Ai() {
	communicator = new SpecificCommunicator(this);
	api = new SpecificApiConnector();
}

Ai::~Ai() {
	delete api;
	delete communicator;
}

void Ai::stop() {
	cout << "VitaminCpp: Stopped." << endl;
}

// User-defined functions
void Ai::initGame(IntMatrix2 architecture,
	IntMatrix2 fruits, IntMatrix2 buildings,
	int limitCherry, int limitKiwi, int limitNut,
	int vitaminGoal, int maxNbTurns) {
	cout << "VitaminCpp: Initialize..." << endl;
	cout << "Goal = " << vitaminGoal << " ; Max turns = " << maxNbTurns << endl;
	
	// Display architecture
	cout << "== Architecture ==" << endl;
	cout << "{" << endl;
	for (int i = 0; i < architecture.size(); i++) {
		cout << "\t{";
		for (int j = 0; j < architecture[0].size(); j++) {
			cout << architecture[i][j] << ", ";
		}
		cout << "}," << endl;
	}
	cout << "}" << endl << endl;
	
	// Display fruits
	cout << "== Fruits ==" << endl;
	for (int i = 0; i < fruits.size(); i++) {
		cout << "Fruit: "
			<< api->decode(fruits[i][OBJECT_TYPE]) << " " << "id #"
			<< fruits[i][OBJECT_ID] << " at (" << fruits[i][OBJECT_X]
			<< ";" << fruits[i][OBJECT_Y] << ")" << endl;
	}
	cout << "Limits: " << limitCherry << " cherries, " << limitKiwi
		<< " kiwies, " << limitNut << " nuts" << endl << endl;
	
	// Display buildings
	cout << "== Buildings ==" << endl;
	for (int i = 0; i < buildings.size(); i++) {
		cout << "Building: "
			<< api->decode(buildings[i][OBJECT_TYPE]) << " " << "id #"
			<< buildings[i][OBJECT_ID] << " at ("
			<< buildings[i][OBJECT_X] << ";" << buildings[i][OBJECT_Y]
			<< ")" << endl;
	}
	cout << endl;
}

void Ai::playTurn(IntMatrix2 newObjects,
	IntMatrix1 deletedObjects, IntMatrix2 movedFruits,
	IntMatrix2 modifiedFruits, IntMatrix2 modifiedSugarDrops) {
	/*cout << "VitaminCpp: Play Turn..." << endl;
	
	// Display new objects
	cout << "== New objects ==" << endl;
	for (int i = 0; i < newObjects.size(); i++) {
		cout << "New object: "
			<< api->decode(newObjects[i][OBJECT_TYPE]) << " " << "id #"
			<< newObjects[i][OBJECT_ID] << " at ("
			<< newObjects[i][OBJECT_X] << ";"
			<< newObjects[i][OBJECT_Y] << ") ; info="
			<< newObjects[i][OBJECT_INFO] << endl;
	}
	cout << endl;
	
	// Display deleted objects
	cout << "== Deleted objects ==" << endl;
	for (int i = 0; i < deletedObjects.size(); i++) {
		cout << "Object: id #" << deletedObjects[i] << " deleted" << endl;
	}
	cout << endl;
	
	// Display moved fruits objects
	cout << "== Moved fruits ==" << endl;
	for (int i = 0; i < movedFruits.size(); i++) {
		cout << "Fruit: id #" << movedFruits[i][OBJECT_ID]
			<< " moved from (" << movedFruits[i][OBJECT_X] << ";"
			<< movedFruits[i][OBJECT_Y] << ") to ("
			<< movedFruits[i][OBJECT_NEW_X] << ";"
			<< movedFruits[i][OBJECT_NEW_Y] << ")" << endl;
	}
	cout <<  endl;
	
	// Display modified fruits objects
	cout << "== Modified fruits ==" << endl;
	for (int i = 0; i < modifiedFruits.size(); i++) {
		cout << "Fruit: id #" << modifiedFruits[i][OBJECT_ID]
			<< " now has " << modifiedFruits[i][OBJECT_LIFE] << " PV and "
			<< modifiedFruits[i][OBJECT_DEFENSE] << " Def" << endl;
	}
	cout << endl;
	
	// Display modified sugar drops objects
	cout << "== Modified sugar drops ==" << endl;
	for (int i = 0; i < modifiedSugarDrops.size(); i++) {
		cout << "Sugar drop: id #"
			<< modifiedSugarDrops[i][OBJECT_ID] << " now has "
			<< modifiedSugarDrops[i][OBJECT_SUGAR] << " sugar grains" << endl;
	}
	cout << endl;
	
	cout << "== Test Game API ==" << endl;
	// Test game API functions
	testGameApi();*/
	/*
	struct timeval start, end;
	long mtime, seconds, useconds;    

	gettimeofday(&start, NULL);
	api->move(1, 42, 42);
	gettimeofday(&end, NULL);

	seconds  = end.tv_sec  - start.tv_sec;
	useconds = end.tv_usec - start.tv_usec;

	mtime = ((seconds) * 1000 + useconds/1000.0) + 0.5;
*/

	std::chrono::time_point<std::chrono::high_resolution_clock> t1 = std::chrono::high_resolution_clock::now();
	api->move(1, 42, 42);
	std::chrono::time_point<std::chrono::high_resolution_clock> t2 = std::chrono::high_resolution_clock::now();

	std::cout << "AI::Move done in " << std::chrono::duration_cast<std::chrono::microseconds > (t2 - t1).count() << "µs" << std::endl;
}

void Ai::mapUpdate(IntMatrix2 newObjects, IntMatrix2 modifiedSugarDrops) {
	/*cout << "VitaminCpp: Map update..." << endl;
	
	// Display new objects
	cout << "== New objects ==" << endl;
	for (int i = 0; i < newObjects.size(); i++) {
		cout << "New object: "
			<< api->decode(newObjects[i][OBJECT_TYPE]) << " " << "id #"
			<< newObjects[i][OBJECT_ID] << " at ("
			<< newObjects[i][OBJECT_X] << ";"
			<< newObjects[i][OBJECT_Y] << ") ; info="
			<< newObjects[i][OBJECT_INFO] << endl;
	}
	cout << endl;
	
	// Display modified sugar drops objects
	cout << "== Modified sugar drops ==" << endl;
	for (int i = 0; i < modifiedSugarDrops.size(); i++) {
		cout << "Sugar drop: id #"
			<< modifiedSugarDrops[i][OBJECT_ID] << " now has "
			<< modifiedSugarDrops[i][OBJECT_SUGAR] << " sugar grains" << endl;
	}
	cout << endl;*/
}

void Ai::testGameApi() {
	cout << "move: " + api->decode(api->move(1, 42, 42)) << endl;
	
	cout << "attack: " + api->decode(api->attack(1, 2)) << endl;
	
	cout << "useEquipment: "
		+ api->decode(api->useEquipment(1, 512, 1)) << endl;
	
	cout << "pickUpEquipment: "
		+ api->decode(api->pickUpEquipment(1, 57)) << endl;
	
	cout << "dropEquipment: "
		+ api->decode(api->dropEquipment(1, 72, 5, 6)) << endl;
	
	cout << "dropSugar: "
		+ api->decode(api->dropSugar(1, 10, 5, 6)) << endl;
	
	cout << "openChest: " + api->decode(api->openChest(1, 18)) << endl;
	
	cout << "stockSugar: " + api->decode(api->stockSugar(1)) << endl;
	
	cout << "sellEquipment: "
		+ api->decode(api->sellEquipment(1, 16)) << endl;
	
	cout << "buyEquipment: "
		+ api->decode(api->buyEquipment(1, EQUIPMENT_SALT_SNIPER)) << endl;
	
	cout << "drinkJuice: " + api->decode(api->drinkJuice(1)) << endl;
	
	cout << "fructify: "
		+ api->decode(api->fructify(FRUIT_CHERRY, 1, 4)) << endl;
	
	cout << "drawVitamin: " + api->decode(api->drawVitamin(1)) << endl;
	
	cout << "drawLine: "
		+ api->decode(api->drawLine(1, 1, 5, 8, COLOR_RED)) << endl;
	
	cout << "drawCircle: "
		+ api->decode(api->drawCircle(8, 9, 2, COLOR_BLUE)) << endl;
}
